src/wasm.h

```diff
@@ -2137,6 +2137,21 @@ struct BinaryLocations {
   std::unordered_map<Function*, FunctionLocations> functions;
 };
 
+// Code annotations for VMs. As with debug info, we do not store these on
+// Expressions as we assume most instances are unannotated, and do not want to
+// add constant memory overhead.
+struct CodeAnnotation {
+  // Branch Hinting proposal: Whether the branch is likely, or unlikely.
+  using BranchLikely = std::optional<bool>;
```

Comment:
I don't think these type aliases are beneficial. Users of `BranchLikely` still need to know that it is actually a `std::optional<bool>`, so hiding that fact from them just introduces more friction. If we do want to introduce a new type name here, using a three-state enum is probably the way to go. Same for `Inline` below.

Reply:
I agree users need to know likely is a bool and inline is an int. But I am trying to help in this way:

```cpp
// ugly "true" - what is it?
makeIf(condition, arm, true);

// comment... slightly better, but easy to forget
makeIf(condition, arm, true /* likely */);

// type - much nicer
makeIf(condition, arm, BranchLikely(true));
```

(Maybe an enforced type would be even better?)

>  If we do want to introduce a new type name here, using a three-state enum is probably the way to go. Same for Inline below

For Inline, a 129-state enum..? :wink:

Reply:
How about using `enum BranchLikeliness { Unlikely, Likely };` and a wrapper type with an explicit constructor:

```
struct InlineHint {
  uint8_t value;
  explicit Inline(uint8_t value) : value(value) {}
);
```

Then for the parameters in IRBuilder use `std::optional<BranchLikeliness>` and `std::optional<InlineHint>`. That puts the optionality where it belongs and makes call sites more readable.

Reply:
I would just continue using `std::optional` explicitly wherever they need to be stored optionally. I generally don't think introducing a new name to avoid writing `std::optional` is a good trade off.

Reply:
Fair enough. In that case, I'm not sure this PR is needed. I'll close it for now - maybe when we start using annotations in more places in the code, we'll have further thoughts on how to improve.